### PR TITLE
[SPARK-5752][SQL] Don't implicitly convert RDDs directly to DataFrames

### PR DESCRIPTION
**Pull request by rxin**
- The old implicit would convert RDDs directly to DataFrames, and that added too many methods. 
- toDataFrame -> toDF
- Dsl -> functions
- implicits moved into SQLContext.implicits
- addColumn -> withColumn
- renameColumn -> withColumnRenamed

Python changes:
- toDataFrame -> toDF
- Dsl -> functions package
- addColumn -> withColumn
- renameColumn -> withColumnRenamed
- add toDF functions to RDD on SQLContext init
- add flatMap to DataFrame

### CODE REVIEW

**Comment by rxin** [comment]
Jenkins, retest this please.

**Comment by rxin** [comment]
Jenkins, retest this please.

**Comment by rxin** [comment]
cc @davies Python test is failing

File: mllib/src/main/scala/org/apache/spark/ml/Transformer.scala [hunk 1]
```diff
@@ -97,7 +97,7 @@ private[ml] abstract class UnaryTransformer[IN, OUT, T <: UnaryTransformer[IN, O
   override def transform(dataset: DataFrame, paramMap: ParamMap): DataFrame = {
     transformSchema(dataset.schema, paramMap, logging = true)
     val map = this.paramMap ++ paramMap
-    dataset.select($"*", callUDF(
-      this.createTransformFunc(map), outputDataType, dataset(map(inputCol))).as(map(outputCol)))
+    dataset.withColumn(map(outputCol),
+      callUDF(this.createTransformFunc(map), outputDataType, dataset(map(inputCol))))
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/Transformer.scala]: minor: use `udf`

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 2]
```diff
@@ -28,17 +29,9 @@ import org.apache.spark.sql.types._
...
-  implicit def symbolToColumn(s: Symbol): ColumnName = new ColumnName(s.name)
-
-  /** Converts $"col name" into an [[Column]]. */
-  implicit class StringToColumn(val sc: StringContext) extends AnyVal {
-    def $(args: Any*): ColumnName = {
-      new ColumnName(sc.s(args :_*))
-    }
-  }
+// scalastyle:off
+object functions {
```
Comment (mengxr) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: minor: `// scalastyle:ignore`

File: sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetTest.scala [hunk 3]
```diff
@@ -90,7 +90,7 @@ trait ParquetTest {
       (f: String => Unit): Unit = {
     import sqlContext.implicits._
     withTempPath { file =>
-      sparkContext.parallelize(data).saveAsParquetFile(file.getCanonicalPath)
+      sparkContext.parallelize(data).toDF().saveAsParquetFile(file.getCanonicalPath)
```
Comment (mengxr) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetTest.scala]: minor: `toDF()` -> `toDF`

**Comment by mengxr** [comment]
LGTM.

**Comment by rxin** [comment]
Jenkins, retest this please.

**Comment by rxin** [comment]
Jenkins, retest this please.

**Comment by rxin** [comment]
Jenkins, retest this please.

**Comment by rxin** [comment]
Jenkins, retest this please.

**Comment by rxin** [comment]
Jenkins, retest this please.

**Pull request closed without merging**